File: BinarySearchTree/BST/BST.cpp
#include "BST.hpp"

/**
 * inserts new integer data into the tree at the position, so that
 * the binary search tree property is maintained.
 * @param data new integer data to be inserted
*/
void BST::insert(int data)
{
    if (root == nullptr) {
        root = new Node(data);
    } else {
        root = insert(root, data);
    }
}

/**
 * insert is an recursive internal method which works on node level
 * @param data new integer data to be inserted
*/
Node* BST::insert(Node* root, int data)
{
    if (root == nullptr) {
        root = new Node(data);
    } else if (data <= root->data) {
        root->left = insert(root->left, data);
    } else {
        root->right = insert(root->right, data);
    }
    return root;
}

/**
 * finds the integer in binary search tree
 * @param data item to be searched
*/
bool BST::search(int data)
{
    if (root == nullptr)
        return false;

    return search(root, data);
}

/**
 * search is an recursive internal method which works at node level
 * @param root root of tree/sub-tre
 * @param data item to be searched
*/
bool BST::search(Node* root, int data)
{
    if (root == nullptr)
        return false;

    if (data == root->data) return true;
    else if (data < root->data) return search(root->left, data);
    else return search(root->right, data);
}

File: golang/bst/bst.go
// Package bst is an implementation for binary search tree data structure
// in go language
package bst

// Node struct contains actual data and address to left and right node
type Node struct {
	Data  int
	Left  *Node
	Right *Node
}

// Tree struct contains the pointer to root node
type Tree struct {
	Root *Node
}

// NewBST creates an empty BST Tree
func NewBST() *Tree {
	return &Tree{nil}
}

// NewNode creates a new Node with integer data
func NewNode(data int) *Node {
	return &Node{data, nil, nil}
}
